loki/expression/mappers.py

```diff
@@ -557,27 +562,46 @@ def map_variable_symbol(self, expr, *args, **kwargs):
             # it does not affect the outcome of expr.clone
             expr.scope.symbol_attrs[expr.name] = expr.type.clone(kind=kind)
 
-        if expr.scope and expr.type.initial and expr.name == expr.type.initial:
-            # FIXME: This is a hack to work around situations where a constant
-            # symbol (from a parent scope) with the same name as the declared
-            # variable is used as initializer. This hands down the correct scope
-            # (in case this traversal is part of ``AttachScopesMapper``) and thus
-            # interrupts an otherwise infinite recursion (see LOKI-52).
+        if kwargs['recurse_to_declaration_attributes']:
             _kwargs = kwargs.copy()
-            _kwargs['scope'] = expr.scope.parent
-            initial = self.rec(expr.type.initial, *args, **_kwargs)
+            _kwargs['recurse_to_declaration_attributes'] = False
+            if expr.scope and expr.type.initial and expr.name == expr.type.initial:
+                # FIXME: This is a hack to work around situations where a constant
+                # symbol (from a parent scope) with the same name as the declared
+                # variable is used as initializer. This hands down the correct scope
+                # (in case this traversal is part of ``AttachScopesMapper``) and thus
+                # interrupts an otherwise infinite recursion (see LOKI-52).
+                _kwargs['scope'] = expr.scope.parent
+                initial = self.rec(expr.type.initial, *args, **_kwargs)
+            else:
+                initial = self.rec(expr.type.initial, *args, **_kwargs)
         else:
-            initial = self.rec(expr.type.initial, *args, **kwargs)
+            initial = expr.type.initial
+
         if initial is not expr.type.initial and expr.scope:
             # Update symbol table entry for initial directly because with a scope attached
             # it does not affect the outcome of expr.clone
             expr.scope.symbol_attrs[expr.name] = expr.type.clone(initial=initial)
 
-        bind_names = self.rec(expr.type.bind_names, *args, **kwargs)
-        if not (bind_names is None or all(new is old for new, old in zip_longest(bind_names, expr.type.bind_names))):
-            # Update symbol table entry for bind_names directly because with a scope attached
-            # it does not affect the outcome of expr.clone
-            expr.scope.symbol_attrs[expr.name] = expr.type.clone(bind_names=as_tuple(bind_names))
+        if kwargs['recurse_to_declaration_attributes']:
+            _kwargs = kwargs.copy()
+            _kwargs['recurse_to_declaration_attributes'] = False
```

Comment:
Good point. I have tried to restructure the control flow in a way that makes it clearer what's going on, and add comments.
In essence: only when recurse_to_declaration_attributes is True, we recurse to all the bits that are stored in a symbol's type. But, because Fortran, we don't want this recursion to happen while actually updating these declaration attributes. Reason is that the declared variable can actually appear in its own initializer expression:
```
REAL var = HUGE(var)
```
Or "it" can appear in an allocation:
```
TYPE(SOME_TYPE), ALLOCATABLE :: var(:)
...
DO j=2,n
    ALLOCATE(var(j)%field(size(var(j-1)%field)))
END DO
```
And since we insert the dimensions from allocate statements as `type.shape`, this now becomes also part of the symbol table entries...

Reply:
Ahhhh, thanks for the explanation. Makes sense and looks much clearer now.